### ISSUE TITLE
shutdown hook NoClassDefFoundError

### ISSUE THREAD

**Issue by randallt** [opentracing-contrib/java-specialagent#118]
Using version 1.3.2, the temporary files are not getting cleaned up and I see this log entry:

2019-07-11 22:24:57,47869 Exception in thread "Thread-1" java.lang.NoClassDefFoundError: io/opentracing/contrib/specialagent/SpecialAgentUtil$2$1
2019-07-11 22:24:57,48324       at io.opentracing.contrib.specialagent.SpecialAgentUtil$2.run(SpecialAgentUtil.java:425)

The class is indeed in the JAR, but apparently it is running under a different classloader?

**Comment by SevaSafris** [comment]
Wow, this is unexpected. Can you provide some more insight with regard to your use-case?

**Comment by randallt** [comment]
I'm trying to use this as part of a custom build. I take it that no one else has seen this? Funny thing is, I haven't changed anything in SpecialAgent before the shutdownhook is registered. The running JVM is using some old framework/library versions. Simple Framework 4.1, Restlet 2.0, as well as older jersey, jetty, and netty versions.

**Comment by SevaSafris** [comment]
We haven’t seen this error before. You’re right to identify the confusing part, that the shutdown hook should have no problem being executed since its anonymous class is constructed when the “shutdown hook is registered”. Unless it’s actually the registration of the shutdown hook that leads to the exception.

How are you integrating SpecialAgent into your build? Are you attempting to have SpecialAgent instrument a SureFire/FailSafe JUnit runtime?

**Comment by randallt** [comment]
I just meant that I'm running a custom build of special agent:
- uses Wavefront as the only a tracer, removes the other tracers since they take up the majority of the JAR size)
- removes some instrumentation rules/plugins to reduce startup time (any other recommendations to reduce startup time? seems to take about 30 sec, which I'm not sure if it will be acceptable to the owners here of the services that I'm trying to instrument. I might need to do some profiling of the startup itself, in which case I will try to communicate my findings back to the project if applicable.)
- adds some custom instrumentation into it that is not applicable to contribute back to the OSS project
- etc.

I'll be working to try to figure this out next, as I already hosed a machine by putting a service into a restart loop that filled the disk. I will communicate my findings here. Any hints, or ideas to try are welcomed.

**Comment by randallt** [comment]
This was solved for me by simply using a concrete inner class for the delete logic. Just need to make sure that no classloading is trying to happen during the shutdown hook. I'm not sure why I was seeing it and no one else is, though. :-(

I also converted it to standard Java 8 syntax, which I think you won't want since you are using Java7.

( SpecialAgentUtil.findJarResources(..) )
```
      ....
      }
      while (resources.hasMoreElements());

      if (destDir != null) {
        final File targetDir = destDir;
        Runtime.getRuntime().addShutdownHook(new DeleteTempFilesShutdownHook(targetDir));
      }
    }
    catch (final IOException e) {
      throw new IllegalStateException(e);
    }
  }

  private static class DeleteTempFilesShutdownHook extends Thread {
	  private final File targetDir;
	  private DeleteTempFilesShutdownHook(File targetDir) {
		  this.targetDir = targetDir;
	  }
	  @Override
	  public void run() {
		  try {
			  Files.walk(targetDir.toPath())
			  	.sorted(Comparator.reverseOrder())
			  	.map(Path::toFile)
			  	.forEach(File::delete);
		  } catch (IOException ioe) {
			  SpecialAgentUtil.logger.severe("Could not delete temporary file: " + ioe.getLocalizedMessage());
		  }
	  }
  }
```

**Comment by randallt** [comment]
Interestingly, the above version also resulted in the exception and leaking temp files occasionally (seems to depend on how long the JVM has been running). Currently trying this version that further forces classloading at registration time:

```
private static class DeleteTempFilesShutdownHook extends Thread {
	  private final Predicate<File> predicate; 
	  private final File targetDir;
	  private DeleteTempFilesShutdownHook(File targetDir) {
		  this.targetDir = targetDir;
		  // force class loading of anonymous inner class when registering
		  this.predicate = new Predicate<File>() {
			  @Override
			  public boolean test(final File t) {
				  return t.delete();
			  }
		  };
	  }
	  @Override
	  public void run() {
		  AssembleUtil.recurseDir(targetDir, predicate);
	  }
  }
```

**Comment by SevaSafris** [comment]
Hi @randallt, thank you for the work you've put to figure this out. I can see how the explicit `DeleteTempFilesShutdownHook` class would solve the `NoClassDefFoundError` you experienced (despite me not being able to reproduce this error.

You have presented an interesting use-case that we did not consider, whereby you desire to remove the Tracer Plugins and Instrumentation Plugins from the build. I will see if this can be implemented directly into the SpecialAgent's build.

Regarding startup time: It is the Instrumentation Plugins that result in a long startup time. You can remove the plugins from inclusion in the SpecialAgent's JAR, or you can disable them from loading with  [this spec](https://github.com/opentracing-contrib/java-specialagent/#34-disabling-instrumentation-plugins). The SpecialAgent has 5 of its own instrumentations that it performs, which means that, even if you remove all Instrumentation Plugins, the SpecialAgent will still require a startup time to instrument the runtime.

The larger the runtime (re the raw byte size of the bytecode of the loaded classes), the longer it will take for the SpecialAgent to start up.

**Comment by randallt** [comment]
Yeah, my lead says we need to get the additional startup time down to only a +15% margin, which I'm not sure is possible. I'll be doing some additional profiling of the startup time within the next couple weeks to see where additional savings can be had.

**Comment by randallt** [comment]
My initial approach to building our agent was going to be to create my own and "borrow" things from this project. However, the complex classloading issues that I almost immediately hit made me re-evaluate. Instead of trying to lift all your solutions to those types of problems, I decided to customize/filter this project itself down to what I needed, and then add the pieces only relevant to us (Wavefront tracer and instrumentation for old libraries). This all took several days to figure out. The Tracer was the trickiest--trying to get maven to put all the right pieces in all the right places in the JAR.

Anything that can be done to make this easier on others moving forward is a great idea.

**Comment by randallt** [comment]
Also, what are your thoughts on instrumenting older libraries, and instrumenting specific versions of libraries in general? Should there be a naming convention in the rules/plugins?

**Comment by SevaSafris** [comment]
Hi @randallt,

Regarding Instrumentation Plugins for old versions of libraries:
Yes, we 100% want to support as many versions as possible. This is especially important for companies with old applications that are running with 10-year-old libraries. In fact, it is more probable that one would be interested in using SpecialAgent for such an application, due to the inability to manually instrument such old code. We ran into this issue with a company that was trying to use SpecialAgent with Spring 3. We quickly implemented Instrumentation Plugins for Spring 3 Web and Spring 3 WebMvc.

Regarding naming supported versions and naming conventions:
The SpecialAgent has a mechanism that allows it to determine whether a plugin is compatible with a 3rd-party library that has invoked the plugin's initialization. In runtime, the version of a 3rd-party library is not available, so the SpecialAgent must use other means to determine compatibility. It uses a class API fingerprinting mechanism to determine if the fingerprint of the 3rd-party library for which the plugin was developed matches the fingerprint of the same 3rd-party library found in the runtime.

As a first guess whether a plugin in SpecialAgent will support a 3rd-party library or not, you can refer to the version numbers next to each plugin mentioned [here](https://github.com/opentracing-contrib/java-specialagent/#61-instrumentation-plugins).

**Comment by randallt** [comment]
Yep, I understand the fingerprinting. I was more asking about adding versions to the rule/plugin names in the specialagent source itself. So, for example, what would multiple versions of Spring Web look like in the 'rules' folder?

**Comment by SevaSafris** [comment]
@randallt, the naming convention of the plugins inside the SpecialAgent don't really matter. These plugins are not themselves externally usable by anything other than SpecialAgent. Therefore, sometime later down the line, we may conceive of a better naming convention. But for the time being, it's not that big of a priority.

**Comment by randallt** [comment]
This has been fixed in release 1.4.0.
